### DISCUSSION TITLE
User permission to only access specific queues in a topic.

### DISCUSSION THREAD

**Discussion by myasul** [rabbitmq/rabbitmq-server#7944] [Questions] [answered]
<img width="1115" alt="image" src="https://user-images.githubusercontent.com/25900564/233246751-1e15dc92-1ad2-484a-aa5d-a4d745b6b243.png">

To set topic permissions, you must use `/api/topic-permissions/vhost/user`. It expects the name of the exchange and regexes to read and write.

What do the regexes match? Does it match the queue name or routing key?

We plan to create users who can only publish and read on specific queues (MQTT). Is that possible with this endpoint? Or is it possible at all?

**Comment by michaelklishin** [comment]
[Access Control](https://rabbitmq.com/access-control.html#authorisation), [Access Control: Topic Authorisation](https://rabbitmq.com/access-control.html#topic-authorisation)

**Comment by michaelklishin** [comment, marked as answer]
All MQTT clients publish to the same topic exchange (`amq.topic` by default but can be overridden).
Topic permissions exist exactly for the needs of MQTT and STOMP users.

The queues used by the MQTT plugin are named using a convention, `mqtt-*`, which allows you to limit access to them for a specific user or set of RabbitMQ users.

More importantly, topic permissions include variable expansion and `client_id` is [one of the expanded variables](https://github.com/rabbitmq/rabbitmq-server/blob/3a17b928686b872cf5920546224530745799de6c/deps/rabbitmq_mqtt/test/java_SUITE.erl#L89), which can help limit access of specific MQTT clients to just their topics.

**Reply by myasul** [reply]
Thank you for answering my question!
